pkg/nimble/rpble/nimble_rpble.c

```diff
@@ -0,0 +1,354 @@
+struct {
+    ble_addr_t addr;        /**< address of highest scored potential parent */
+    uint16_t score;         /* 0 := not used, larger is better! */
+} _psel = { { 0 }, SCORE_NONE };
+
+/* eval event used for periodical state updates */
+static uint32_t _eval_itvl;
+static struct ble_npl_callout _evt_eval;
+
+static nimble_netif_eventcb_t _eventcb = NULL;
+
+static uint16_t _psel_score(uint16_t rank, uint8_t free)
+{
+    return (UINT16_MAX - rank + free);
+}
+
+static void _children_accept(void)
+{
+    int res;
+    (void)res;
+
+    /* stop advertising while we are updating the context */
+    nimble_netif_accept_stop();
+
+    /* we only start advertising (accepting) if we do have an active parent and
+     * if we have resources for new connections */
+    if ((_current_parent == PARENT_NONE) ||
+        (nimble_netif_conn_count(NIMBLE_NETIF_UNUSED) == 0)) {
+        return;
+    }
+
+    /* compile RPL meta data into AD service data TLV field */
+    uint8_t sd[AD_SVC_DATA_LEN];
+    byteorder_htolebufs(sd, BLE_GATT_SVC_IPSS);
+    sd[POS_INST_ID] = _local_rpl_ctx.inst_id;
+    memcpy(&sd[POS_DODAG_ID], _local_rpl_ctx.dodag_id, 16);
+    sd[POS_VERSION] = _local_rpl_ctx.version;
+    byteorder_htolebufs(&sd[POS_RANK], _local_rpl_ctx.rank);
+    sd[POS_FREE_SLOTS] = (uint8_t)nimble_netif_conn_count(NIMBLE_NETIF_UNUSED);
+
+    /* generate the new advertising data */
+    bluetil_ad_t ad;
+    uint8_t buf[BLE_HS_ADV_MAX_SZ];
+    res = bluetil_ad_init_with_flags(&ad, buf, BLE_HS_ADV_MAX_SZ,
+                                     BLUETIL_AD_FLAGS_DEFAULT);
+    assert(res == BLUETIL_AD_OK);
+    res = bluetil_ad_add(&ad, BLE_GAP_AD_SERVICE_DATA, sd, sizeof(sd));
+    assert(res == BLUETIL_AD_OK);
+
+    /* start advertising this node */
+    res = nimble_netif_accept(ad.buf, ad.pos, &_adv_params);
+    assert(res == NIMBLE_NETIF_OK);
+}
+
+static void _on_scan_evt(uint8_t type,
+                         const ble_addr_t *addr, int8_t rssi,
+                         const uint8_t *ad, size_t ad_len)
+{
+    int res;
+
+    /* filter out all non-connectible advertisements */
+    if (type != BLE_HCI_ADV_TYPE_ADV_IND) {
+        return;
+    }
+
+    /* check if scanned node does actually speak rpble */
+    bluetil_ad_data_t sd_field;
+    bluetil_ad_t ads = { .buf = (uint8_t *)ad, .pos = ad_len, .size = ad_len };
+    res = bluetil_ad_find(&ads, BLE_GAP_AD_SERVICE_DATA, &sd_field);
+    if (res != BLUETIL_AD_OK) {
+        return;
+    }
+    if ((sd_field.len != AD_SVC_DATA_LEN) ||
+        (byteorder_lebuftohs(sd_field.data) != BLE_GATT_SVC_IPSS)) {
+        return;
+    }
+
+    /**
+     * @note    Here we need to improve the filtering: so far, we consider every
+     *          node we see that is capable of rplbe to be a parent. We should
+     *          however also filter for instance ID and possibly the DODAG ID as
+     *          well. On top, we should probably only consider nodes with >=
+     *          version as parent
```

Comment:
What's stopping us from doing just that?

Reply:
so far only the additional time needed to add this. At some point I decided to PR the implementation state I have. Its still useful and it can always be enhanced once in, right?!